docs/USAGE-GUIDE.md

```diff
@@ -1373,6 +1382,71 @@ int s2n_config_add_ticket_crypto_key(struct s2n_config *config, const uint8_t *n
+**op** object representing this operation. From this function the user
+can spawn the thread with **op** object to perform through
+**s2n_async_pkey_perform_op** call and immediately return from the
+function without waiting for thread to complete. The **s2n_negotate**
+will fail with **S2N_ERR_T_BLOCKED** error type and **s2n_blocked_status**
+**S2N_BLOCKED_ON_APPLICATION_INPUT**, and will keep giving the same
+error until the **op** is performed on the spawned thread and consumed
+by connection. Once the spawned thread finishes performing the **op**,
+the **s2n_connection_consume_async_op** needs to be called in the main
+event loop thread for connection to consume it. After that the user needs
+to call **s2n_negotiate** to proceed with handshake.
+
+```c
+typedef int (*s2n_async_pkey_fn)(struct s2n_connection *conn, struct s2n_async_pkey_op *op);
+extern int s2n_config_set_async_pkey_callback(struct s2n_config *config, s2n_async_pkey_fn fn);
+extern int s2n_async_pkey_perform_op(struct s2n_async_pkey_op *op, s2n_cert_private_key *key);
+extern int s2n_connection_consume_async_op(struct s2n_connection *conn, struct s2n_async_pkey_op *op);
+extern int s2n_async_pkey_op_free(struct s2n_async_pkey_op *op);
+```
+
+- **op** is an opaque object representing private key operation which
+needs to be performed.
+- **key** is a private key used for operation.
+
+**s2n_async_pkey_fn** is invoked every time some action involving
+private key is required. The **conn** provides a pointer to connection
+which triggered this action. The **op** is an operation which needs to be
+performed through **s2n_async_pkey_perform_op** call and consumed by
+**conn** through **s2n_connection_consume_async_op** call after the
+callback was complete.
+
+**s2n_config_set_async_pkey_callback** sets up the callback to invoke
+for asynchronous private key operations and enables asynchronous mode.
+When it is used **s2n_negotiate** will error out with **S2N_ERR_T_BLOCKED**
+error type and **s2n_blocked_status** **S2N_BLOCKED_ON_APPLICATION_INPUT**
+each time the callback was invoked and won't proceed until
+**s2n_connection_consume_async_op** is called on the connection and
+associated **op** object.
+
+**s2n_async_pkey_perform_op** performs the **op** allowing it to be used
+to resume the handshake through **s2n_connection_consume_async_op**
+call. This function can be called only once and is safe to call from a
+different thread. To get the private key **key** the caller can use
+**s2n_connection_get_selected_cert** and **s2n_cert_chain_and_key_get_key**
+calls.
+
+**s2n_connection_consume_async_op** will consume the **op** after it was
+performed through **s2n_async_pkey_perform_op** call. It will release
+the memory for **op** if call succeeds. The caller will need to call
+**s2n_negotiate** again in order to proceed with TLS handshake.
+
+**s2n_async_pkey_op_free** releases the **op**. Should not be used if
```

Comment:
Is it safe to be called before `s2n_async_pkey_perform_op` was executed?

And do you have to `free` the op with and without having called `s2n_async_pkey_perform_op`?

Reply:
It is safe to call anytime except for after **s2n_connection_consume_async_op**, which frees the **op** for you. Otherwise you can manually free it at any time, but this means you won't be able to perform the handshake. Basically it only should be used if something went wrong and you want to release the memory for **op** and likely **conn** as well.

Reply:
I think I asked the same thing again on the new diff 😂

Understood now - but I think the docs should include what you mentioned here.

Reply:
Yeah, will do - didn't get to updating the docs yet (:

Reply:
Updated